internal/git/provider/provider.go

```diff
@@ -2,20 +2,75 @@ package provider
 
 import (
 	"fmt"
+	"net/url"
+	"strings"
 
+	"github.com/go-logr/logr"
 	"golang.org/x/net/context"
 )
 
+const GitHubProviderName = "github"
+
 type Provider interface {
-	ListPullRequests(ctx context.Context, repo Repository) []PullRequest
-	AddCommentToPullREquest(ctx context.Context, repo PullRequest, comment []byte)
+	ListPullRequests(ctx context.Context, repo Repository) ([]PullRequest, error)
+	AddCommentToPullRequest(ctx context.Context, repo PullRequest, body []byte) (*Comment, error)
+
+	SetLogger(logr.Logger) error
+	SetToken(tokenType, token string) error
+	SetHostname(hostname string) error
+
+	Setup() error
 }
 
-func New(provider string) (Provider, error) {
+func New(provider string, options ...ProviderOption) (Provider, error) {
+	var p Provider
 	switch provider {
-	case "github":
-		return newGitHubProvider(), nil
+	case GitHubProviderName:
+		p = newGitHubProvider()
 	default:
 		return nil, fmt.Errorf("unknown provider: %s", provider)
 	}
+
+	for _, opt := range options {
+		if err := opt(p); err != nil {
+			return nil, err
+		}
+	}
+
+	return p, nil
+}
+
+func FromURL(repoURL string, options ...ProviderOption) (Provider, Repository, error) {
+	targetProvider := ""
+	repo := Repository{}
+
+	parsedURL, err := url.Parse(repoURL)
+	if err != nil {
+		return nil, repo, fmt.Errorf("unable to parse url: %w", err)
+	}
+
+	// That's a pretty naiv check for now.
+	// Use proper parsing later with a well tested library.
```

Comment:
Let's add a new issue for this so that we don't forget 👍

Reply:
https://github.com/weaveworks/tf-controller/issues/669